builtins: add crdb_internal.read/write_file

dt opened the pull request:
Release note (security update): Added the admin-only debugging functions
crdb_internal.read/write_file to read/write bytes from/to external storage URIs
used by BACKUP or IMPORT.

dt committed:
builtins: add crdb_internal.read/write_file

Release note (security update): Added the admin-only debugging functions
crdb_internal.read/write_file to read/write bytes from/to external storage URIs
used by BACKUP or IMPORT.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67427)
<!-- Reviewable:end -->

pbardea commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -3429,6 +3429,41 @@ may increase either contention or retry errors, or both.`,
+				}
+				return tree.NewDInt(tree.DInt(len(data))), nil
+			},
+			Info:       "Read the content of a file at an external storage URI",
```
This needs updating for Write rather than read

pbardea reviewed (approved):
LGTM with minor doc comment

knz reviewed (changes requested):
huh admin only perhaps?

dt commented:
@knz heh, yep, that's why I added the blocking label and added reviewer yesterday: going to discuss it next week to determine if we should match `IMPORT` -- admin-only for reads to arbitrary endpoints or using host auth -- or just blanket admin-only.

dt commented:
Actually, on second though, this is just a debugging tool -- just always admin-only seems fine rather than bike shedding something more granular/complex.

Updated and RFAL.

knz reviewed (commented):
LGTM, but please add a release note with tag "security update" to report that these new admin-only functions can access/write files out of the process.

> Reviewed 1 of 5 files at r1, 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67427#-:-MeZwcdN1hekeYrX8Z0h:bify0ec)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aaron-crl, @ajwerner, and @dt)



<!-- Sent from Reviewable.io -->

dt commented:
@knz does the fact that an admin could already use IMPORT to read  arbitrary remote files, or use EXPORT to write them, affect your determination as to if we need the extra warning?

dt commented:
TFTRs!

bors r+
